src/data/data.cc

```diff
@@ -40,10 +40,35 @@ void MetaInfo::Clear() {
 }
 
 void MetaInfo::SaveBinary(dmlc::Stream *fo) const {
+  /**
+   * Header: version, num_field, followed by list of form [ (field_name, field_offset) ].
+   * Offset is to be calculated from the end of the header.
+   **/
   Version::Save(fo);
-  fo->Write(&num_row_, sizeof(num_row_));
-  fo->Write(&num_col_, sizeof(num_col_));
-  fo->Write(&num_nonzero_, sizeof(num_nonzero_));
+  const uint64_t num_field = kNumField;
+  fo->Write(num_field);
+  uint64_t offset = 0;
+  int field_cnt = 0;  // make sure we are actually writing kNumField fields
+  fo->Write(std::string(u8"num_row")); fo->Write(offset);
+  offset += sizeof(num_row_); ++field_cnt;
+  fo->Write(std::string(u8"num_col")); fo->Write(offset);
+  offset += sizeof(num_col_); ++field_cnt;
+  fo->Write(std::string(u8"num_nonzero")); fo->Write(offset);
```

Comment:
Just my personal opinion, I prefer using raw string (char*) with dmlc stream, as it has specialization for `std::string` that comes with corresponding read method, sometimes it's really difficult to reason about the binary offset with dmlc stream.  I bumped into it while refactoring model IO.

Reply:
My understanding is that dmlc stream saves the length of the string followed by the string content. The length field is always 8 bytes long.